Document new API for Node analyzers (#6252)
opensearch-project/documentation-website: _analyzers/normalizers.md

Consider the following example.

Create a new index with custom normalizer:
```json
PUT /sample-index
{
  "settings": {
    "analysis": {
      "normalizer": {
        "normalized_keyword": {
          "type": "custom",
          "char_filter": [],
          "filter": [ "asciifolding", "lowercase" ]
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "approach": {
        "type": "keyword",
        "normalizer": "normalized_keyword"
      }
    }
  }
}
```
{% include copy-curl.html %}

Index document:
```json
POST /sample-index/_doc/
{
  "approach": "naive"
}
```
{% include copy-curl.html %}

The following query matches the document. This is expected:
```json
GET /sample-index/_search
{
  "query": {
    "term": {
      "approach": "naive"
    }
  }
}
```
{% include copy-curl.html %}

But this query matches the document as well:
```json
GET /sample-index/_search
{
  "query": {
    "term": {
      "approach": "Naïve"
    }
  }
}
```
{% include copy-curl.html %}

> To understand why, try to see the effect of the normalizer:

Reviewer: ```suggestion
To understand why, consider the effect of the normalizer:
```

```json
GET /sample-index/_analyze
{
  "normalizer" : "normalized_keyword",
  "text" : "Naïve"
}
```

> Internally, a normalizer accepts only filters that are instances of either `NormalizingTokenFilterFactory` or `NormalizingCharFilterFactory`. The following is a list of compatible filters found in modules and plugins that are part of the core OpenSearch repository:

Reviewer: ```suggestion
Internally, a normalizer accepts only filters that are instances of either `NormalizingTokenFilterFactory` or `NormalizingCharFilterFactory`. The following is a list of compatible filters found in modules and plugins that are part of the core OpenSearch repository.
```

> ### Module `common-analysis`

Reviewer 2: ```suggestion
### The `common-analysis` module
```

> ### Module `common-analysis`

Reviewer: @kolchfa-aws As with the "Step" headings in an earlier PR, the word order of this and the following three headings is confusing. Please revise.

Author: @kolchfa-aws I think we can resolve this one too now.

> This module does not require installation, it is available by default.

Reviewer: ```suggestion
This module does not require installation; it is available by default.
```

Character filters: `pattern_replace`, `mapping`

Token filters: `arabic_normalization`, `asciifolding`, `bengali_normalization`, `cjk_width`, `decimal_digit`, `elision`, `german_normalization`, `hindi_normalization`, `indic_normalization`, `lowercase`, `persian_normalization`, `scandinavian_folding`, `scandinavian_normalization`, `serbian_normalization`, `sorani_normalization`, `trim`, `uppercase`

> ### Plugin `analysis-icu`

Reviewer 2: ```suggestion
### The `analysis-icu` plugin
```

Character filters: `icu_normalizer`

> Token filter: `icu_normalizer`, `icu_folding`, `icu_transform`

Reviewer: ```suggestion
Token filters: `icu_normalizer`, `icu_folding`, `icu_transform`
```

> ### Plugin `analysis-kuromoji`

Reviewer 2: ```suggestion
### The `analysis-kuromoji` plugin
```

Character filters: `normalize_kanji`, `normalize_kana`

> ### Plugin `analysis-nori`

Reviewer 2: ```suggestion
### The `analysis-nori` plugin
```

Character filters: `normalize_kanji`, `normalize_kana`

> This list of filters covers only analysis components found in [additional plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/#additional-plugins) that are part of the core OpenSearch repository.

Reviewer: ```suggestion
These lists of filters include only analysis components found in the [additional plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/#additional-plugins) that are part of the core OpenSearch repository.
```

---
layout: default
title: Normalizers
nav_order: 100
---

# Normalizers

> A Normalizer is similar to Analyzer except it does not contain tokenizer, and it can only include specific type of Character filters and Token filters. These filters can do only character level operations, such as character or pattern replacement, but can not operate on the whole token level, this means that replacing the token with synonym or stemming is not supported.

Author: Here I am going to propose changes in the beginning of the sentence. Please check when the PR is updated.

> A Normalizer is similar to Analyzer except it does not contain tokenizer, and it can only include specific type of Character filters and Token filters. These filters can do only character level operations, such as character or pattern replacement, but can not operate on the whole token level, this means that replacing the token with synonym or stemming is not supported.

Author: FYI: PR is updated.

GET /sample-index/_analyze
{
  "normalizer" : "normalized_keyword",
  "text" : "Naïve"
}
```

Internally, Normalizer accept only filters that are instances of either `NormalizingTokenFilterFactory` or `NormalizingCharFilterFactory`. The following is a list of compatible filters found in modules and plugins that are part of core OpenSearch repository:

### Module `common-analysis`

This module does not require installation, it is available by default.

Character filters: `pattern_replace`, `mapping`

Token filters: `arabic_normalization`, `asciifolding`, `bengali_normalization`, `cjk_width`, `decimal_digit`, `elision`, `german_normalization`, `hindi_normalization`, `indic_normalization`, `lowercase`, `persian_normalization`, `scandinavian_folding`, `scandinavian_normalization`, `serbian_normalization`, `sorani_normalization`, `trim`, `uppercase`

### Plugin `analysis-icu`

Character filters: `icu_normalizer`

Token filter: `icu_normalizer`, `icu_folding`, `icu_transform`

### Plugin `analysis-kuromoji`

Character filters: `normalize_kanji`, `normalize_kana`

### Plugin `analysis-nori`

Character filters: `normalize_kanji`, `normalize_kana`

> This list of filters covers only analysis components found in [additional plugins]({{{site.url}}{{site.baseurl}}/install-and-configure/plugins/#additional-plugins) that are part of the core OpenSearch repository.

Author: Good catch!
